libraries/AP_Compass/AP_Compass_AK09916.cpp

```diff
@@ -66,175 +93,347 @@ AP_Compass_Backend *AP_Compass_AK09916::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice>
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Unable to get bus semaphore\n");
         return false;
     }
 
-    if (bus_type == AK09916_ICM20948_I2C && dev_icm) {
-        uint8_t rval;
-        if (!dev_icm->read_registers(REG_ICM_WHOAMI, &rval, 1) ||
-            rval != ICM_WHOAMI_VAL) {
-            // not an ICM_WHOAMI
-            goto fail;
-        }
-        uint8_t retries = 5;
-        do {
-            // reset then bring sensor out of sleep mode
-            if (!dev_icm->write_register(REG_ICM_PWR_MGMT_1, 0x80)) {
-                goto fail;
-            }
-            hal.scheduler->delay(10);
-
-            if (!dev_icm->write_register(REG_ICM_PWR_MGMT_1, 0x00)) {
-                goto fail;
-            }
-            hal.scheduler->delay(10);
-            
-            // see if ICM20948 is sleeping
-            if (!dev_icm->read_registers(REG_ICM_PWR_MGMT_1, &rval, 1)) {
-                goto fail;
-            }
-            if ((rval & 0x40) == 0) {
-                break;
-            }
-        } while (retries--);
-       
-        if (rval & 0x40) {
-            // it didn't come out of sleep
-            goto fail;
-        }
+    if (!_bus->configure()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Could not configure the bus\n");
+        goto fail;
+    }
 
-        // initially force i2c bypass off
-        dev_icm->write_register(REG_ICM_INT_PIN_CFG, 0x00);
-        hal.scheduler->delay(1);
-
-        // now check if a AK09916 shows up on the bus. If it does then
-        // we have both a real AK09916 and a ICM20948 with an embedded
-        // AK09916. In that case we will fail the driver load and use
-        // the main AK09916 driver
-        uint16_t whoami;
-        if (dev->read_registers(REG_COMPANY_ID, (uint8_t *)&whoami, 2)) {
-            // a device is replying on the AK09916 I2C address, don't
-            // load the ICM20948
-            hal.console->printf("ICM20948: AK09916 bus conflict\n");
-            goto fail;
-        }
+    if (!_reset()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Reset Failed\n");
+        goto fail;
+    }
 
-        // now force bypass on
-        dev_icm->write_register(REG_ICM_INT_PIN_CFG, 0x02);
-        hal.scheduler->delay(1);
+    if (!_check_id()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Wrong id\n");
+        goto fail;
     }
 
-    uint16_t whoami;
-    if (!dev->read_registers(REG_COMPANY_ID, (uint8_t *)&whoami, 2) ||
-        whoami != 0x0948) {
-        // not a AK09916
+    if (!_setup_mode()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Could not setup mode\n");
         goto fail;
     }
 
-    dev->setup_checked_registers(2);
+    if (!_bus->start_measurements()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Could not start measurements\n");
+        goto fail;
+    }
 
-    dev->write_register(REG_CNTL2, 0x08, true); // continuous 100Hz
```

Comment:
are we no longer using continuous mode?

Reply:
Yes we are, its just mode setup is done from:
https://github.com/ArduPilot/ardupilot/blob/9f50feadfed3e0066e5c5fcc0db147fcb6abcac4/libraries/AP_Compass/AP_Compass_AK09916.cpp#L308-L310